File: swift/Sources/AetherProtocol/Routing/RouteStore.swift
// SPDX-License-Identifier: MIT

import Foundation

/// Persistent backing store for the routing table. Default is in-memory.
public protocol RouteStore: Sendable {
    func get(_ destinationUhid: String) async -> RouteEntry?
    func getAll() async -> [RouteEntry]
    func save(_ route: RouteEntry) async
    func remove(_ destinationUhid: String) async
    func pruneExpired() async -> Int
}

/// Process-local route store. Loses everything on restart.
public actor InMemoryRouteStore: RouteStore {
    private var routes: [String: RouteEntry] = [:]

    public init() {}

    public func get(_ destinationUhid: String) -> RouteEntry? { routes[destinationUhid] }
    public func getAll() -> [RouteEntry] { Array(routes.values) }
    public func save(_ route: RouteEntry) { routes[route.destination] = route }
    public func remove(_ destinationUhid: String) { routes.removeValue(forKey: destinationUhid) }

    public func pruneExpired() -> Int {
        let expiredKeys = routes.filter { $0.value.isExpired }.map { $0.key }
        for k in expiredKeys { routes.removeValue(forKey: k) }
        return expiredKeys.count
    }
}

File: kotlin/src/main/kotlin/aethermesh/routing/RouteStore.kt
// SPDX-License-Identifier: MIT

package aethermesh.routing

import aethermesh.models.RouteEntry
import java.util.concurrent.ConcurrentHashMap

/**
 * Persistent backing store for the routing table. Default implementation is in-memory;
 * production hosts substitute file- or SQLite-backed implementations for durability.
 */
interface RouteStore {
    suspend fun get(destinationUhid: String): RouteEntry?
    suspend fun getAll(): List<RouteEntry>
    suspend fun save(route: RouteEntry)
    suspend fun remove(destinationUhid: String)
    suspend fun pruneExpired(): Int
}

/** Process-local route store. Loses everything on restart. */
class InMemoryRouteStore : RouteStore {
    private val routes = ConcurrentHashMap<String, RouteEntry>()

    override suspend fun get(destinationUhid: String): RouteEntry? = routes[destinationUhid]

    override suspend fun getAll(): List<RouteEntry> = routes.values.toList()

    override suspend fun save(route: RouteEntry) {
        routes[route.destinationUhid] = route
    }

    override suspend fun remove(destinationUhid: String) {
        routes.remove(destinationUhid)
    }

    override suspend fun pruneExpired(): Int {
        var pruned = 0
        val expired = routes.entries.filter { it.value.isExpired() }
        for ((k, _) in expired) {
            if (routes.remove(k) != null) pruned++
        }
        return pruned
    }
}
